### PR TITLE
feat: Add optionnal detailed data per Label

### PR DESCRIPTION
**Pull request by frco9**
Add an option to get a more detailed response with data per Labels which can be useful to compute metrics per Labels as well as the global ones

### CODE REVIEW

**Comment by piercus** [comment]
Bonjour Jérémie, (@frco9 )

Thanks a lot, I'm glad to see this PR ! 
Code and Pull Requests looks good to me, but it would be great if you can add a simple unit tests for this in `test./map.js` and in `test/list-misclassified.js` 

Merci beaucoup

**Comment by frco9** [comment]
Hi @piercus,

I've added some test and fixed a typo ! 
Tell me if you want me to add more test cases

File: test/list-misclassified.js [hunk 1]
```diff
@@ -69,3 +69,77 @@ test('basic example', t => {
...
+	}];
+
+	const res = listMisclassified({
+		groundTruths,
+		predictions,
+		perLabelDetail: true
+	});
+
+	t.true(Array.isArray(res.perLabels));
+	t.true(res.perLabels.length === 2);
```
Comment (piercus) [inline test/list-misclassified.js]:
```suggestion
	t.true(res.perLabel.length === 2);
```

File: test/list-misclassified.js [hunk 2]
```diff
@@ -69,3 +69,77 @@ test('basic example', t => {
...
+	t.true(res.perLabels[0].label === 'car');
```
Comment (piercus) [inline test/list-misclassified.js]:
```suggestion
	t.true(res.perLabel[0].label === 'car');
```

File: test/list-misclassified.js [hunk 3]
```diff
@@ -69,3 +69,77 @@ test('basic example', t => {
...
+	t.true(res.perLabels[1].label === 'pedestrian');
```
Comment (piercus) [inline test/list-misclassified.js]:
```suggestion
	t.true(res.perLabel[1].label === 'pedestrian');
```

File: test/list-misclassified.js [hunk 4]
```diff
@@ -69,3 +69,77 @@ test('basic example', t => {
 		bottom: 92
 	}]);
 });
+
+test('test case: if perLabelDetail defined should return object with all misclassified object and an array of misclassified per labels ', t => {
```
Comment (piercus) [inline test/list-misclassified.js]:
```suggestion
test('test case: if perLabel defined should return object with all misclassified object and an array of misclassified per labels ', t => {
```

File: test/map.js [hunk 5]
```diff
@@ -198,3 +198,46 @@ test('test case: if iouThreshold defined ', t => {
 	t.true(res >= 0);
 	t.true(res <= 1);
 });
+
+test('test case: if perLabelDetail defined should return details per labels ', t => {
```
Comment (piercus) [inline test/map.js]:
```suggestion
test('test case: if perLabel defined should return details per labels ', t => {
```

File: test/map.js [hunk 6]
```diff
@@ -198,3 +198,46 @@ test('test case: if iouThreshold defined ', t => {
...
+	}];
+
+	const iouThreshold = 0.6;
+	const res = mAP({
+		groundTruths,
+		predictions,
+		iouThreshold,
+		perLabelDetail: true
+	});
+	t.true(Array.isArray(res.aPs));
```
Comment (piercus) [inline test/map.js]:
```suggestion
	t.true(Array.isArray(res.aps));
```

File: test/map.js [hunk 7]
```diff
@@ -198,3 +198,46 @@ test('test case: if iouThreshold defined ', t => {
...
+	t.true(res.aPs.length === 2);
```
Comment (piercus) [inline test/map.js]:
```suggestion
	t.true(res.aps.length === 2);
```

File: test/map.js [hunk 8]
```diff
@@ -198,3 +198,46 @@ test('test case: if iouThreshold defined ', t => {
...
+	t.true(res.aPs[0].label === 'car');
```
Comment (piercus) [inline test/map.js]:
```suggestion
	t.true(res.aps[0].label === 'car');
```

File: test/map.js [hunk 9]
```diff
@@ -198,3 +198,46 @@ test('test case: if iouThreshold defined ', t => {
...
+	t.true(res.aPs[1].label === 'pedestrian');
```
Comment (piercus) [inline test/map.js]:
```suggestion
	t.true(res.aps[1].label === 'pedestrian');
```

File: lib/map.js [hunk 10]
```diff
@@ -43,9 +44,13 @@ module.exports = function ({
...
 
-		return ap({groundTruthsLabel, predictionsLabel, iouThreshold});
+		return {label: l, ap: ap({groundTruthsLabel, predictionsLabel, iouThreshold})};
 	});
 
 	const l = aps.length;
-	return aps.reduce((a, b) => a + b, 0) / l;
+	const mAp = aps.reduce((a, b) => a + b.ap, 0) / l;
+	if (perLabelDetail) {
+		return {mAp, aPs: aps};
```
Comment (piercus) [inline lib/map.js]: let's use `mAP` instead of `mAp`

```suggestion
		return {mAP, aPs: aps};
```

File: lib/map.js [hunk 11]
```diff
@@ -43,9 +44,13 @@ module.exports = function ({
 		const groundTruthsLabel = groundTruths.filter(({label}) => label === l).map(o => Object.assign({}, o, {match: false}));
 		const predictionsLabel = predictions.filter(({label}) => label === l).map(o => Object.assign({}, o, {match: false}));
 
-		return ap({groundTruthsLabel, predictionsLabel, iouThreshold});
+		return {label: l, ap: ap({groundTruthsLabel, predictionsLabel, iouThreshold})};
 	});
 
 	const l = aps.length;
-	return aps.reduce((a, b) => a + b, 0) / l;
+	const mAp = aps.reduce((a, b) => a + b.ap, 0) / l;
```
Comment (piercus) [inline lib/map.js]:
```suggestion
	const mAP = aps.reduce((a, b) => a + b.ap, 0) / l;
```

**Comment by frco9** [comment]
Thanks @piercus, I’ll do the changes on Monday ! Have a nice weekend

File: lib/map.js (hunk 10 above)
Comment (frco9) [reply lib/map.js]: I guess you've changed your mind ? Because in the email notification I got, you suggested to use `mAp` and `aps` !?

**Comment by frco9** [comment]
@piercus I've changed all your reviews, and also renamed the `aPs` to `aps` it looked weird

**Comment by piercus** [comment]
Thanks, I will do full validation on Wednesday

**Comment by piercus** [comment]
Thanks @frco9 for your help on this !
It's good for me.

Eventually, if you want to add an other PR with a little example in the README.md i would accept it also :smiley:

**Pull request merged**